### PR TITLE
Always use service id as block id

### PR DESCRIPTION
**Pull request by core23**
This is the BC version for #302

### Changelog

``` markdown
### Changed
- empty block names are automatically set via `TweakCompilerPass`
```
### Subject

The block id is automatically set to the internal service id (if empty). This is the same mechanism like creating a admin.

### CODE REVIEW

**Comment by greg0ire** [comment]
Why do we want that again? I read #302, but it's unclear to me.

File: DependencyInjection/Compiler/TweakCompilerPass.php [hunk 1]
```diff
@@ -31,12 +31,31 @@ public function process(ContainerBuilder $container)
...
+                // NEXT_MAJOR: Remove deprecation notice
+                @trigger_error(
+                    'Using different service id and block id is deprecated since 3.x.',
+                    E_USER_DEPRECATED
+                );
+            }
+
+            if (strlen($arguments[0]) == 0) {
+                $definition->replaceArgument(0, $id);
+            }
```
Comment (greg0ire) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: Uuuuuh… again :open_mouth: ?

File: DependencyInjection/Compiler/TweakCompilerPass.php [hunk 2]
```diff
@@ -31,12 +31,31 @@ public function process(ContainerBuilder $container)
 
         $parameters = $container->getParameter('sonata_block.blocks');
 
-        foreach ($container->findTaggedServiceIds('sonata.block') as $id => $attributes) {
+        foreach ($container->findTaggedServiceIds('sonata.block') as $id => $tags) {
```
Comment (greg0ire) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: `$tags` is not used, you could write `array_keys($container->findTaggedServiceIds('sonata.block'))` instead, might be less confusing

File: UPGRADE-3.x.md [hunk 3]
```diff
@@ -1,6 +1,8 @@
 UPGRADE 3.x
 ===========
 
+Injecting the block id is deprecated and will be automatically set.
```
Comment (greg0ire) [inline UPGRADE-3.x.md]: What should the user do though? Leave an empty `<argument/>` ?

If yes please add :

```
Instead, provide an empty (null?) argument:

## Before
`<argument>your.block</argument>`

## After
`<argument/>`
```

File: DependencyInjection/Compiler/TweakCompilerPass.php (hunk 2 above)
Comment (core23) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: Working with `array_keys` here looks more confusing to me

File: DependencyInjection/Compiler/TweakCompilerPass.php (hunk 1 above)
Comment (core23) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: Copy & paste problem ;)

**Comment by core23** [comment]
> Why do we want that again? I read #302, but it's unclear to me.

ATM it's a little confusing, that we have two different ids for a `BlockService`, the service id and the block id. We should change that part, so that the id's are corresponding. We use the same mechanism for [admin services ](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L50-L58)(admin id = service id).

File: DependencyInjection/Compiler/TweakCompilerPass.php (hunk 2 above)
Comment (greg0ire) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: Ok, it must be a matter of taste, so do as you wish. Be aware that `phpmd` will whine about this kind of thing though.

**Comment by greg0ire** [comment]
Try amend + force push, for StyleCI?

File: DependencyInjection/Compiler/TweakCompilerPass.php (hunk 2 above)
Comment (core23) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: That's not the only code where `phpmd` will whine ;)

**Comment by greg0ire** [comment]
:confused: Maybe contact their support (there is an email on the StyleCI page)?

**Comment by greg0ire** [comment]
Can you rebase @core23 ? I fixed the Travis build : https://github.com/sonata-project/SonataBlockBundle/pull/345/files#diff-0e4afb0bb3966e755707777f8ec7f823L4

**Comment by core23** [comment]
Rebased

**Comment by greg0ire** [comment]
@sonata-project/contributors please review

File: DependencyInjection/Compiler/TweakCompilerPass.php [hunk 4]
```diff
@@ -31,12 +31,27 @@ public function process(ContainerBuilder $container)
...
+
+            $arguments = $definition->getArguments();
+
+            // Replace empty block id with service id
+            if (strlen($arguments[0]) == 0) {
+                $definition->replaceArgument(0, $id);
+            } elseif ($id != $arguments[0]) {
+                // NEXT_MAJOR: Remove deprecation notice
+                @trigger_error(
+                    'Using different service id and block id is deprecated since 3.x.',
```
Comment (OskarStark) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: And will be removed in 4.0

Comment (core23) [inline DependencyInjection/Compiler/TweakCompilerPass.php]: Fixed that too

**Comment by OskarStark** [comment]
Thank you @core23 👍

**Comment by Jean85** [comment]
This BC/deprecation is missing from the changelog and from the upgrade guide.
I would write a PR myself, but I'm not understanding how to fix this...

**Comment by core23** [comment]
That's a known problem, because we can't rename this services without a BC break.

**Pull request merged**